### ISSUE TITLE
Can we use `\Drupal\Core\Field\WidgetInterface::isApplicable` instead of an alter hook?

### ISSUE THREAD

**Issue by deviantintegral** [juanolalla/readonly_field_widget#4]
`\Drupal\Core\Field\WidgetInterface::isApplicable` reads like it could be used in place of the alter hook, which probably be clearer as it would keep all of the code in the same class.

**Comment by juanolalla** [comment]
Done: https://github.com/juanolalla/readonly_field_widget/commit/b4b1bb862097d44329fec2bff6a497c003a60a95

**Comment by juanolalla** [comment]
I don't think we can use this. That static function is called for every field of the entity (even for status, or created), a few times for each of them in each request. And not only per request, it is called also via ajax for each field every time a user chooses a widget, even if it is not the chosen widget in that field select input.

The hook is just about changing the widget information, such as file types. It is called just once, and you would need to clear cache to call it again and apply changes. I don't know any other way to do just that, except in the widget annotation, but I'm using the hook because we need a dynamic set field types, not static.

So following your idea of clarity, I've moved the logic of getting and filtering the field types in a static function in the class: https://github.com/juanolalla/readonly_field_widget/commit/7fa09a0ecf740d8cc1d94846133d35c88fcb074f

**Comment by juanolalla** [comment]
Quick fix: https://github.com/juanolalla/readonly_field_widget/commit/9a73dce8138f4365e4aa82a9cf99a7c5283ead0b

**Comment by deviantintegral** [comment]
Perfect, thanks for checking into this.
